File: aplicação/backend/src/main/java/com/cloudfeedback/service/FeedbackService.java
package com.cloudfeedback.service;

import com.cloudfeedback.model.Feedback;
import java.util.UUID;
import com.amazonaws.services.dynamodbv2.AmazonDynamoDB;
import com.amazonaws.services.dynamodbv2.AmazonDynamoDBClientBuilder;
import com.amazonaws.services.dynamodbv2.document.DynamoDB;
import com.amazonaws.services.dynamodbv2.document.Table;
import com.amazonaws.services.dynamodbv2.document.Item;
import com.amazonaws.services.sns.AmazonSNS;
import com.amazonaws.services.sns.AmazonSNSClientBuilder;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class FeedbackService {

    private static final Logger logger = LoggerFactory.getLogger(FeedbackService.class);

    private final DynamoDB dynamoDB;
    private final Table feedbackTable;
    private final AmazonSNS snsClient;
    private final String snsTopicArn = "arn:aws:sns:sa-east-1:120726684266:feedback-topic";

    public FeedbackService() {
        AmazonDynamoDB client = AmazonDynamoDBClientBuilder.standard().build();
        this.dynamoDB = new DynamoDB(client);
        this.feedbackTable = dynamoDB.getTable("Feedbacks");
        this.snsClient = AmazonSNSClientBuilder.standard().build();
    }

    public void processFeedback(Feedback feedback) {
        try {
            // Gerar UUID
            feedback.setFeedbackId(UUID.randomUUID().toString());

            // Preencher dataEnvio se vier nulo
            if (feedback.getDataEnvio() == null || feedback.getDataEnvio().isBlank()) {
                String agora = LocalDateTime.now().format(DateTimeFormatter.ISO_DATE_TIME);
                feedback.setDataEnvio(agora);
            }

            // Salvar no DynamoDB
            feedbackTable.putItem(new Item()
                    .withPrimaryKey("feedback_id", feedback.getFeedbackId())
                    .withString("nome", feedback.getNome())
                    .withString("email", feedback.getEmail())
                    .withString("mensagem", feedback.getMensagem())
                    .withString("data_envio", feedback.getDataEnvio())
            );

            // Enviar notificação SNS
            String mensagem = "Novo feedback de " + feedback.getNome() + ": " + feedback.getMensagem();
            snsClient.publish(snsTopicArn, mensagem, "Novo feedback recebido");

        } catch (Exception e) {
            // Logar erro para CloudWatch
            logger.error("Erro ao processar feedback: {}", e.getMessage(), e);
            // Aqui você pode lançar uma RuntimeException para que a Lambda retorne 500
            throw new RuntimeException("Falha ao processar feedback", e);
        }
    }
}

File: aplicação/backend/service/feedback_service.py
import uuid
from datetime import datetime
import boto3
import logging

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

class FeedbackService:
    def __init__(self):
        self.dynamodb = boto3.resource('dynamodb')
        self.table = self.dynamodb.Table('Feedbacks')
        self.sns = boto3.client('sns')
        self.sns_topic_arn = 'arn:aws:sns:sa-east-1:120726684266:feedback-topic'

    def process_feedback(self, feedback):
        try:
            feedback_id = feedback.feedback_id or str(uuid.uuid4())
            data_envio = feedback.data_envio or datetime.now().isoformat()
            logger.info(f"Inserindo feedback no DynamoDB: {feedback_id} | Nome: {feedback.nome}")
            self.table.put_item(Item={
                'feedback_id': feedback_id,
                'nome': feedback.nome,
                'email': feedback.email,
                'mensagem': feedback.mensagem,
                'data_envio': data_envio
            })
            assunto = "Novo feedback recebido"
            mensagem_sns = f"Novo feedback de {feedback.nome}:\n\n{feedback.mensagem}"
            logger.info(f"Publicando feedback no SNS: {self.sns_topic_arn}")
            response = self.sns.publish(
                TopicArn=self.sns_topic_arn,
                Message=mensagem_sns,
                Subject=assunto
            )
            logger.info(f"SNS publish response: {response}")
        except Exception as e:
            logger.error(f"Erro ao processar feedback: {e}", exc_info=True)
            raise
